Issue #3810: Replaced singletonTag in nested html rules in javadoc grammar with singletonElement

ps-sp opened the pull request:
Issue #3810 : Replaced singletonTag in nested html rules in javadoc grammar with singletonElement

ps-sp commented:
[antlr diff report](https://3810-antlrdiff.netlify.com/checkstyle/)

Vladlis commented:
@ps-sp 
PR looks good, please rebase onto the latest master and update Parser/Lexer coverage

ps-sp commented:
@Vladlis 

>PR looks good, please rebase onto the latest master and update Parser/Lexer coverage

Updated pom.xml

ps-sp commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:
```diff
@@ -44,7 +44,7 @@
  * bottom) of all the imports, or above (or under) each group, or are treated
  * like non static imports (@see {@link ImportOrderOption}</li>
- * </ul>
+ * </ul>.
```
To make SummaryJavadoc pass

Vladlis replied:
Please explain how is this related.

ps-sp replied:
CIs were red. Checkstyle inspection for checkstyle files was failing. `SummaryJavadoc` was failing. `SummaryJavadoc` [has ties with](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java#L175) `htmlElement` rule. Structure for this file changed with the update to grammar and `SummaryJavadoc` was detecting missed period for javadoc paragraph. I will upload regression reports for `SummaryJavadoc`.

rnveach replied:
Since this is becomming connected to `SummaryJavadoc`, please add a UT to `SummaryJavadoc` so we can have some documentation of this.

@ps-sp Missing a change in grammar almost got by unnoticed on changes in checks. Nothing picked this up but this one little change and it wasn't obvious at all.
I suggest you do regression on all javadoc checks with any ANTLR changes from now on to make sure this doesn't happen anywhere else.

romani commented:
@ps-sp , please rebase

ps-sp commented:
[diff report](http://3810-javadoc-diff.getforge.io/)

rnveach commented:
http://3810-javadoc-diff.getforge.io/hibernate-orm/index.html#A7
http://3810-javadoc-diff.getforge.io/sevntu-checkstyle/index.html#A1
http://3810-javadoc-diff.getforge.io/Orekit/index.html#A1
Change in `JavadocParagraph`.
http://3810-javadoc-diff.getforge.io/checkstyle/index.html#A1
http://3810-javadoc-diff.getforge.io/apache-struts/index.html#A14
http://3810-javadoc-diff.getforge.io/apache-struts/index.html#A26
Change in `SummaryJavadoc`.

@ps-sp We have also now expanded into `JavadocParagraphCheck`. Please pick some unique examples and add them to the UTs.

romani commented:
Please rebase

Vladlis commented:
@ps-sp 
Please also take https://github.com/checkstyle/checkstyle/issues/3570 into account

ps-sp commented:
[Javadoc checks diff report](https://3810-javadocchecks.netlify.com/)

ps-sp commented:
@rnveach @Vladlis 
Change in `JavadocParagraph` relates to #3570 and is related to [JavadocParagraphCheck.java#L131-#L133](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L131-#L133)

Changing those lines to 

```
else if (ast.getType() == JavadocTokenTypes.HTML_ELEMENT) {
            int firstChildType = JavadocUtils.getFirstChild(ast).getType();
            if (firstChildType == JavadocTokenTypes.P_TAG_START
                    || firstChildType == JavadocTokenTypes.PARAGRAPH) {
                checkParagraphTag(ast);
            }
        }
```

solves 3570 and should make changes in `JavadocParagraph` go away.

BUT right now `JavadocParagraph` check isn't designed to process "nested" `paragraph`s apparently and I don't get the exact coding style philosphy with which it was done in the first place, but was it intentional ? Because making the above fix results in more than 1000 errors from `JavadocParagraph` on checkstyle files all concerned about `<p>` tag's position.

AND if it is intentional to not process nested paragraphs in `JavadocParagraph` then contrary to what @rnveach said in #3570 issue description 

>[Reading the description](http://checkstyle.sourceforge.net/config_javadoc.html#JavadocParagraph), the violations look valid and should stay regardless of the <br/> or not.

Those violations should **NOT** stay regardless for `<br/>` because in the input case `<p>` tag is supposed to be nested.

ps-sp commented:
@rnveach @Vladlis 
Also I added an UT for `SummaryJavadoc` and updated PR. The UT is supposed to reflect the reason because of which modification to `ImportOrderCheck` was needed in this PR. `SummaryJavadoc` is supposed to grab all the `text` it finds at the first level of javadoc tree to create the [summaryDoc](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java#L138) variable. What's happening in master is that `summaryDoc` ends up getting text from within `<p>` tag which didn't get nested because of having a `singletonElement` within it and thus the `text` inside `<p>` tag ended up at level 1 and eventually got grabbed inside `summaryDoc` at [SummaryJavadocCheck.java#L190](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java#L190) and [coincidentally that text had a period at the end.](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L155)

Vladlis commented:
> BUT right now JavadocParagraph check isn't designed to process "nested" paragraphs apparently and I don't get the exact coding style philosphy with which it was done in the first place, but was it intentional ? 

I think it is just a bug and violations have to be solved in scope of #3570

ps-sp commented:
[performance report](https://3810-performance.netlify.com/)

ps-sp commented:
[performance report](https://3810-performance.netlify.com/)

Vladlis commented:
@rnveach @romani , Please review

rnveach commented:
https://github.com/checkstyle/checkstyle/pull/4787#issuecomment-318538506
> Change in JavadocParagraph.
> Change in SummaryJavadoc.

I still see no UTs under these specific checks.
I expect changes in `JavadocParagraphCheckTest` and `SummaryJavadocCheckTest`.

ps-sp commented:
> I still see no UTs under these specific checks

A case has been added for `SummaryJavadoc`. Please refer to this comment [4787#issuecomment](https://github.com/checkstyle/checkstyle/pull/4787#issuecomment-321358352)

ps-sp commented:
@rnveach 
As far as `JavadocParagraph` is concerned please have a look at [4787#issuecomment](https://github.com/checkstyle/checkstyle/pull/4787#issuecomment-321354253)

rnveach commented:
> A case has been added for SummaryJavadoc. 

I see it now.

> JavadocParagraph 

I don't know the check well enough to say, but UT should still be added since it is a change in it's behavior basically.
@romani Please respond to reasoning and logic of check from @ps-sp 's comment.

ps-sp commented:
> I don't know the check well enough to say, but UT should still be added since it is a change in it's behavior basically.

That UT might end up `assert`ing wrong behaviour, since there are some problems with the check itself. The check logic should be changed in #3570 is what I felt. Of course the check logic can't be changed here and a UT should only be added after the check logic is corrected. Why add a wrong UT here ? We can just make a note of this and do all the related work in scope of #3570 ?
